Extract and improve script to download CUDA toolkit components.

stellaraccident opened the pull request:
Previous to this, we were fetching a sample from an NVIDIA github repo and using CMake scripting to use it to download an appropriate SDK. This patch:

* Forks the parse_redist.py sample locally into third_party/nvidia_sdk_download.
* Fixes a number of things in parse_redist.py to make it more robust, remove warnings, and eliminate the dependency on the Python requests package.
* Removes the 'requests' package from setup.py releases files as it is no longer needed. Leaves it in runtime/build_requirements.txt as it is still used in some benchmark scripts.
* Adds a fetch_cuda_toolkit.py which duplicates the behavior that was open coded in CMake scripting.
* Updates the build_tools/third_party/cuda/CMakeLists.txt to use the new script instead of its internal approach.

In a follow-on, I will use this script on the Bazel side to make it auto-fetch the CUDA SDK as needed as well.

Commit:
Extract and improve script to download CUDA toolkit components.

Previous to this, we were fetching a sample from an NVIDIA github repo and using CMake scripting to use it to download an appropriate SDK. This patch:

* Forks the parse_redist.py sample locally into third_party/nvidia_sdk_download.
* Fixes a number of things in parse_redist.py to make it more robust, remove warnings, and eliminate the dependency on the Python requests package.
* Removes the 'requests' package from all requirements files as it is no longer needed.
* Adds a fetch_cuda_toolkit.py which duplicates the behavior that was open coded in CMake scripting.
* Updates the build_tools/third_party/cuda/CMakeLists.txt to use the new script instead of its internal approach.

In a follow-on, I will use this script on the Bazel side to make it auto-fetch the CUDA SDK as needed as well.

ScottTodd commented on third_party/nvidia_sdk_download/parse_redist.py:
```diff
@@ -0,0 +1,302 @@
+the http_get helper) by using urllib directly
+  - explicit error handling on hash mismatch
+  - always download, even if files exist
+  - remove dependence on deprecated distutils copy_tree in favor of shutil
```
It would be helpful to structure this PR as two commits (and then merge without squashing): one that adds a pristine copy of this file from that link, then a second commit that applies the listed changes.

stellaraccident replied:
I can land it that way. Please proceed with reviewing as if done.

ScottTodd replied:
Two commits is also easier to review (saves a few steps of opening the original file and manually diffing, when GitHub's UI can shows diffs between two commits already)

stellaraccident replied:
Split.

ScottTodd commented on build_tools/third_party/cuda/CMakeLists.txt:
```diff
@@ -5,79 +5,18 @@
+  )
+  if(_EXEC_RESULT AND NOT _EXEC_RESULT EQUAL 0)
+    message(FATAL_ERROR "Error fetching CUDA toolkit")
   endif()
```
The first message here looks inaccurate? (It's not _just_ checking, it's actually performing a download)

ScottTodd commented on runtime/pyproject.toml:
```diff
@@ -7,6 +7,5 @@ requires = [
     "packaging",
     "pybind11>=2.10.1",
     "PyYAML",
-    "requests",
```
Something in here is still using this:

https://github.com/openxla/iree/actions/runs/4370994923/jobs/7646455762#step:7:250
```
 82/110 Test #115: iree/build_tools/benchmarks/post_benchmark_comment_test ................................***Failed    0.18 sec
Traceback (most recent call last):
  File "C:\a\iree\iree\build_tools\benchmarks\post_benchmark_comment_test.py", line 9, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
```

stellaraccident replied:
Adding back in.

Commit:
Add pristine copy of NVIDIA parse_redist.py and LICENSE.

Forked from: https://raw.githubusercontent.com/NVIDIA/build-system-archive-import-examples/355e25cca11725e88984443a6a343dffeb43308a/parse_redist.py

Applied project yapf formatting to pristine copy.

GMNGeoffrey commented on build_tools/third_party/cuda/CMakeLists.txt:
```diff
@@ -5,79 +5,18 @@
+    OUTPUT_VARIABLE _ACTUAL_DOWNLOAD_PATH
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+  if(_EXEC_RESULT AND NOT _EXEC_RESULT EQUAL 0)
```
What if no `_EXEC_RESULT`? :fearful:

GMNGeoffrey commented on third_party/nvidia_sdk_download/fetch_cuda_toolkit.py:
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+# Copyright 2022 The IREE Authors
```
Wait now we have a thing under the IREE license that's in a directory with a different license... :see_no_evil:

This is just shelling out to parse_redist, so it seems not that hard to locate it elsewhere and find the file (separately it would be nice to not have to invoke another process there, but I think it's not worth rewriting the script to avoid it)
